modulemd_tools/modulemd_tools/yaml.py

```diff
@@ -0,0 +1,300 @@
+    # for buildrequires.
+    old_deps = Modulemd.Dependencies()
+
+    # Module can contain multiple pairs of dependencies. If we want to update
+    # both `requires` and `buildrequires` at the same time, we can drop all
+    # current dependencies and just set a new one. If we want to update only
+    # one of them, we are getting to an ambiguous situation, not knowing what
+    # pair of dependencies we should update. Let's just raise an exception.
+    if (len(mod_stream.get_dependencies()) > 1
+            and (requires or buildrequires)
+            and not (requires and buildrequires)):
+        raise AttributeError("Provided YAML contains multiple pairs of "
+                             "dependencies. It is ambiguous which one to "
+                             "update.")
+
+    if mod_stream.get_dependencies():
+        old_deps = mod_stream.get_dependencies()[0]
+    new_deps = Modulemd.Dependencies()
+    if requires:
+        for depname, depstreams in requires.items():
+            for depstream in depstreams:
+                new_deps.add_runtime_stream(depname, depstream)
+    else:
+        for depname in old_deps.get_runtime_modules():
+            for depstream in old_deps.get_runtime_streams(depname):
+                new_deps.add_runtime_stream(depname, depstream)
+
+    if buildrequires:
+        for depname, depstreams in buildrequires.items():
+            for depstream in depstreams:
+                new_deps.add_buildtime_stream(depname, depstream)
+    else:
+        for depname in old_deps.get_buildtime_modules():
+            for depstream in old_deps.get_buildtime_streams(depname):
+                new_deps.add_buildtime_stream(depname, depstream)
+
+    if requires or buildrequires:
+        mod_stream.clear_dependencies()
+        mod_stream.add_dependencies(new_deps)
+
+    idx2 = Modulemd.ModuleIndex.new()
+    idx2.add_module_stream(mod_stream)
+    return idx2.dump_to_string()
+
+
+def upgrade(mod_yaml, version):
+    """
+    Upgrade the input module YAML from an older version to a newer one.
+    Downgrades aren't supported even in case where it would be possible.
+    """
+    parsed = yaml.safe_load(mod_yaml or "")
+    if not parsed or not "version" in parsed:
+        raise ValueError("Missing modulemd version")
+
+    supported = [1, 2]
+    if parsed["version"] not in supported or version not in supported:
+        raise ValueError("Unexpected modulemd version")
+
+    if parsed["version"] > version:
+        raise ValueError("Cannot downgrade modulemd version")
+
+    mod_stream = Modulemd.ModuleStream.new(parsed["version"])
+    mod_stream = mod_stream.read_string(mod_yaml, True,
+                                        parsed["data"].get("name", ""),
+                                        parsed["data"].get("stream", ""))
+    mod_stream_upgraded = mod_stream.upgrade(version)
+    return _stream2yaml(mod_stream_upgraded)
```

Comment:
This entire function is unnecessary, since everything it is doing is already done by the `ModuleStream.upgrade()` call.